### PR TITLE
Refactor ReduceMean/Sum Gradient without Shape Dependency.

### PR DESCRIPTION
**Pull request by iK1D**
Refactor the ReduceMean and ReduceSum gradient without using the input/output shapes.

**Motivation and Context**
- Recent optimization for “removing redundant computation” at the last layer for BERT and UniLM v2 introduces NonZero Ops, which generate dynamic output shapes. If the loss Op doesn't reduce, but use reduce_mean or reduce_sum after that, current gradient implementation will not work because the it depends on the shapes which are not available. This PR refactors the ReduceMean and ReduceSum gradient without using the shapes.

### CODE REVIEW

File: orttraining/orttraining/test/gradient/gradient_ops_test.cc [hunk 1]
```diff
@@ -403,7 +403,8 @@ TEST(GradientCheckerTest, GemmGrad) {
 TEST(GradientCheckerTest, ReduceMeanGrad) {
   float max_error;
   GradientChecker<float, float, float> gradient_checker;
-  OpDef op_def{"ReduceMean"};
+  // Attribute axes supports negative values from opset 11.
+  OpDef op_def{"ReduceMean", kOnnxDomain, 11};
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/gradient/gradient_ops_test.cc:407]: Can we add a test case for negative axes values?

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 2]
```diff
@@ -746,120 +746,64 @@ IMPLEMENT_GRADIENT_BUILDER(GetDivGradient) {
...
+  result.push_back(NodeDef("Size", {I(0)}, {IA("Scale_Denominator")}));
   result.push_back(
-      NodeDef("Tile",
-              {unsqueezed_Grad, REPEATS},
-              {IA("Tiled_Grad", IType(0))}));
-
-  NodeDef scale_node = ConstantValueNode(1.0f / static_cast<float>(scale), Name("Scale"));
-  ArgDef SCALE = scale_node.output_args[0];
-  result.push_back(scale_node);
+      NodeDef("Cast",
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/gradient_builder.cc:767]: Are these cast really needed? 
Div also support int64_t input type

Comment (SherlockNoMad) [reply orttraining/orttraining/core/graph/gradient_builder.cc:767]: maybe it's the output of Div need a cast?

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 3]
```diff
@@ -746,120 +746,64 @@ IMPLEMENT_GRADIENT_BUILDER(GetDivGradient) {
...
+              {IA("Casted_Scale_Numerator")},
+              {MakeAttribute("to", type_float)}));
+  result.push_back(
+      NodeDef("Div",
+              {IA("Casted_Scale_Numerator"), IA("Casted_Scale_Denominator")},
+              {IA("Scale")}));
+  result.push_back(NodeDef("Mul", {grad, IA("Scale")}, {IA("Scaled_Grad")}));
+  result.push_back(NodeDef("Shape", {I(0)}, {IA("Shaped_X")}));
+  result.push_back(NodeDef("ConstantOfShape", {IA("Shaped_X")}, {IA("Constant_Of_Shaped_X")}));
+  result.push_back(NodeDef("Add", {IA("Scaled_Grad"), IA("Constant_Of_Shaped_X")}, {GI(0)}));
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/gradient_builder.cc:784]: This Add is still a work around after all... 
I think we should introduce a "Broadcast" contrib op, inputs are X and target_shape. X repeats itself until the target shape. 
I have seen this op in pytorch and caffe2 and feel that it will be a useful op for building training graph.

Comment (SherlockNoMad) [reply orttraining/orttraining/core/graph/gradient_builder.cc:784]: Performance-wise, it will also be better than using Add.

File: orttraining/orttraining/core/graph/gradient_builder.cc (hunk 2 above)
Comment (iK1D) [reply orttraining/orttraining/core/graph/gradient_builder.cc:767]: Here numerator <= denominator so the Div result is 0<r<=1, Div on integer won't work.

File: orttraining/orttraining/core/graph/gradient_builder.cc (hunk 3 above)
Comment (iK1D) [reply orttraining/orttraining/core/graph/gradient_builder.cc]: On 2nd thought, Expand is exactly what we want, it's same as numpy's broadcast_to. I also applied an optimization here as what I did for the Where kernel, which will improve the compute time a lot (I don't have an exact percentage number for now, but I did observe the difference when I played with the UniLM v2 model by adding expand to the model and making this kernel change).

File: orttraining/orttraining/core/graph/gradient_builder.cc (hunk 2 above)
Comment (SherlockNoMad) [reply orttraining/orttraining/core/graph/gradient_builder.cc:767]: Make sense.

File: orttraining/orttraining/core/graph/gradient_builder.cc (hunk 3 above)
Comment (SherlockNoMad) [reply orttraining/orttraining/core/graph/gradient_builder.cc]: Yes!! Expand is what we are looking for.
